File: gameengine/src/network/Packet.cpp
#include "Packet.h"

#include <stdexcept>
#include <string.h>

#include "serialiser/Serialise.h"
#include "serialiser/Deserialise.h"

std::vector<Serialiser::Field> Packet::GetFields(Type type)
{
	if (type == Type::ConnEstablished)
	{
		return std::vector<Serialiser::Field>({
			Serialiser::Field(Serialiser::Type::Int32, offsetof(ConnEstablished, uid))
			});
	}
	else if (type == Type::ClientChatMessage)
	{
		return std::vector<Serialiser::Field>({
			Serialiser::Field(Serialiser::Type::UnlimitedString, offsetof(ClientChatMessage, message))
			});
	}
	else if (type == Type::ServerChatMessage)
	{
		return std::vector<Serialiser::Field>({
			Serialiser::Field(Serialiser::Type::NullTerminatedString, offsetof(ServerChatMessage, name)),
			Serialiser::Field(Serialiser::Type::UnlimitedString, offsetof(ServerChatMessage, message))
			});
	}
	else if (type == Type::SetClientName)
	{
		return std::vector<Serialiser::Field>({
			Serialiser::Field(Serialiser::Type::UnlimitedString, offsetof(SetClientName, name))
			});
	}
	else if (type == Type::SetScene)
	{
		return std::vector<Serialiser::Field>({
			Serialiser::Field(Serialiser::Type::NullTerminatedString, offsetof(SetScene, root)),
			Serialiser::Field(Serialiser::Type::UnlimitedString, offsetof(SetScene, file))
			});
	}
	else
	{
		throw std::invalid_argument("Unknown type: " + std::to_string(static_cast<int>(type)));
	}
}

std::vector<Serialiser::Field> Packet::GetFields() const
{
	return this->GetFields(this->GetType());
}

Packet::Packet(std::vector<char> bytes)
{
	this->Deserialise(bytes);
}

void Packet::SetType(Type type)
{
	if (type == Type::ConnEstablished)
	{
		this->m_data = ConnEstablished();
	}
	else if (type == Type::ClientChatMessage)
	{
		this->m_data = ClientChatMessage();
	}
	else if (type == Type::ServerChatMessage)
	{
		this->m_data = ServerChatMessage();
	}
	else if (type == Type::SetClientName)
	{
		this->m_data = SetClientName();
	}
	else
	{
		throw std::invalid_argument("Unknown type: " + std::to_string(static_cast<int>(type)));
	}
}

Packet::Type Packet::GetType() const
{
	return static_cast<Packet::Type>(this->m_data.index());
}

std::vector<char> Packet::Serialise() const
{
	std::vector<char> result;
	result.push_back(static_cast<char>(this->GetType()));

	std::vector<char> body_bytes;
	std::vector<Serialiser::Field> fields = this->GetFields();
	if (this->GetType() == Type::ConnEstablished)
	{
		body_bytes = Serialiser::Serialise(this->GetData<ConnEstablished>(), fields);
	}
	else if (this->GetType() == Type::ClientChatMessage)
	{
		body_bytes = Serialiser::Serialise(this->GetData<ClientChatMessage>(), fields);
	}
	else if (this->GetType() == Type::ServerChatMessage)
	{
		body_bytes = Serialiser::Serialise(this->GetData<ServerChatMessage>(), fields);
	}
	else if (this->GetType() == Type::SetClientName)
	{
		body_bytes = Serialiser::Serialise(this->GetData<SetClientName>(), fields);
	}
	else
	{
		throw std::invalid_argument("Unknown type: " + std::to_string(static_cast<int>(this->GetType())));
	}

	result.insert(result.end(), body_bytes.begin(), body_bytes.end());

	return result;
}

void Packet::Deserialise(std::vector<char> bytes)
{
	if (bytes.size() == 0)
	{
		throw std::invalid_argument("No bytes provided");
	}
	else
	{
		this->SetType(static_cast<Type>(bytes.at(0)));

		std::vector<Serialiser::Field> fields = this->GetFields();
		if (this->GetType() == Type::ConnEstablished)
		{
			this->m_data = Serialiser::Deserialise<ConnEstablished>(bytes, fields, 1);
		}
		else if (this->GetType() == Type::ClientChatMessage)
		{
			this->m_data = Serialiser::Deserialise<ClientChatMessage>(bytes, fields, 1);
		}
		else if (this->GetType() == Type::ServerChatMessage)
		{
			this->m_data = Serialiser::Deserialise<ServerChatMessage>(bytes, fields, 1);
		}
		else if (this->GetType() == Type::SetClientName)
		{
			this->m_data = Serialiser::Deserialise<SetClientName>(bytes, fields, 1);
		}
		else
		{
			throw std::invalid_argument("Unknown packet header");
		}
	}
}

File: gameengine-server/src/GameEngineServer/Network/Packet.hs
module GameEngineServer.Network.Packet (Packet (..), serialise, deserialise, PacketType (..), packetToPacketType) where

import qualified Data.ByteString
import qualified Data.ByteString.Lazy
import Data.ByteString.Char8 (pack, unpack, readInteger)
import Data.ByteString.Builder (toLazyByteString, Builder, string8, word8, int32LE, char8)
import Data.Word (Word8)
import Data.Int (Int32)
import Data.Char (chr)
import Data.Binary.Get (Get, runGet, getWord8, getInt32le, getRemainingLazyByteString)

{-
Checklist for adding new Packet types:
1. Add to Packet
2. Add to instance Show Packet
3. Add to serialiseInner
4. Add to getPacket
5. Add to PacketType
6. Add to packetToPacketType
-}

-- |Type used to represent information being sent between the server and the client
data Packet =
    -- |Sent from server to client containing the unique ID of that client when the client connects
    ConnEstablished Integer
    -- |Sent by either the client containing global chat messages
    | ClientChatMessage String
    -- |Sent by the server containing chat messages and the user who sent them
    | ServerChatMessage String String
    -- |Sent by a client containing its new name
    | SetClientName String
    -- |Sent by the server containing the new root and relative path to the scene file
    | SetScene String String

instance Show Packet where
    show (ConnEstablished uid) = "ConnEstablished {uid: " ++ show uid ++ "}"
    show (ClientChatMessage message) = "ClientChatMessage {message: \"" ++ message ++ "\"}"
    show (ServerChatMessage name message) = "ChatMessage {name: \"" ++ name ++ "\", " ++ "message: \"" ++ message ++ "\"}"
    show (SetClientName name) = "SetClientName {name: \"" ++ name ++ "\"}"
    show (SetScene root relative) = "SetClientName {root: \"" ++ root ++ "\", relative: \"" ++ relative ++ "\"}"

-- |Turn a 'Packet' into a 'ByteString' with the correct header and delimiter ready to be broadcasted
serialise :: Packet -> Data.ByteString.ByteString
serialise packet = if body_length > (2 ^ 8) then error "Body must contain less than 256 bytes" else (Data.ByteString.Lazy.toStrict . Data.ByteString.Lazy.concat) [header, body]
    where
        body = toLazyByteString (mappend ((word8 . packetToNum) packet) (serialiseInner packet))
        body_length = Data.ByteString.Lazy.length body
        header = (toLazyByteString . word8) (fromIntegral body_length)

serialiseInner :: Packet -> Builder
serialiseInner (ConnEstablished uid) = mconcat [int32LE (fromIntegral uid)]
serialiseInner (ClientChatMessage _) = error "Server can't ClientChatMessage"
serialiseInner (ServerChatMessage name message) = mconcat [string8 name, char8 '\0', string8 message]
serialiseInner (SetClientName _) = error "Server can't send SetClientName"
serialiseInner (SetScene root relative) = mconcat [string8 root, char8 '\0', string8 relative]

-- |Turn a 'ByteString' into a 'Packet' if it conforms to the correct layout
deserialise :: Data.ByteString.ByteString -> Packet
deserialise bytes = runGet getPacket (Data.ByteString.Lazy.fromStrict bytes)

-- |Get a 'Packet' from a lazy 'ByteString' using 'runGet'
getPacket :: Get Packet
getPacket = do
    packetType <- getWord8
    case numToPacketType packetType of
        TypeClientChatMessage -> do
            message <- getRemainingString
            return (ClientChatMessage message)
        
        TypeSetClientName -> do
            message <- getRemainingString
            return (SetClientName message)
    
        _ -> error ("Unknown header type received - " ++ show packetType)

-- |Get the remaining 'Char's and pack into a 'String' from a lazy 'ByteString' using 'runGet'
getRemainingString :: Get String
getRemainingString = fmap (unpack . Data.ByteString.Lazy.toStrict) getRemainingLazyByteString

-- |Enumerators representing constructors for 'Packet'
data PacketType = TypeConnEstablished | TypeClientChatMessage | TypeServerChatMessage | TypeSetClientName | TypeSetScene deriving (Show, Eq, Enum)

-- |Convert a 'Packet' to a 'PacketType' enumerator
packetToPacketType :: Packet -> PacketType
packetToPacketType (ConnEstablished _) = TypeConnEstablished
packetToPacketType (ClientChatMessage _) = TypeClientChatMessage
packetToPacketType (ServerChatMessage _ _) = TypeServerChatMessage
packetToPacketType (SetClientName _) = TypeSetClientName
packetToPacketType (SetScene _ _) = TypeSetScene

-- |Convert a 'PacketType' to some 'Num' that can be used when serialising and deserialising
packetTypeToNum :: Num a => PacketType -> a
packetTypeToNum packetType = fromIntegral $ fromEnum packetType

-- |Convert a 'Packet' to some 'Num' that can be used when serialising and deserialising
packetToNum :: Num a => Packet -> a
packetToNum = packetTypeToNum . packetToPacketType

-- |Convert 'Integral' (i.e. 'Word8') at the start of a raw 'Packet' into a 'PacketType'
numToPacketType :: Integral a => a -> PacketType
numToPacketType n = toEnum $ fromIntegral n